File: EstruturasFlexiveis/OutrasEstruturas/Queue/Java/Queue.java
package EstruturasFlexiveis.OutrasEstruturas.Queue.Java;

public class Queue<T> {
    //=====PRIVATE=====//
    private Node<T> first;
    private Node<T> last;
    private int length = 0;

    //=====CONSTRUCTORS=====//
    public Queue() {
        this.setFirst(null);
        this.setLast(null);
    }

    @SuppressWarnings("unchecked")
    public Queue(T... arr) {
        for(T obj : arr) 
            enqueue(obj);
    }

    //=====GET=====//
    private Node<T> getFirst() {
        return this.first;
    }

    private Node<T> getLast() {
        return this.last;
    }

    public int size() {
        return this.length;
    }
    
    //=====SET=====//
    private void setFirst(Node<T> first) {
        this.first = first;
    }

    private void setLast(Node<T> last) {
        this.last = last;
    }

    private void inc() {
        this.length++;
    }

    private void dec() {
        this.length--;
    }

    //=====METODOS=====//
    public Boolean isEmpty() {
        return this.size() == 0;
    }
    
    //=====ENQUEUE=====//
    public void enqueue(T obj) {
        Node<T> node = new Node<>(obj, null);
        
        if(isEmpty()) {
            this.setFirst(node);
            this.setLast(node);
        } else {
            this.getLast().setNext(node);
            this.setLast(node);
        }

        inc();
    }

    //=====DEQUEUE=====//
    public T dequeue() {
        if(isEmpty()) throw new EmptyQueueException();

        T content = this.getFirst().getObj();
        if(this.getFirst() == this.getLast()) {
            this.setFirst(null);
            this.setLast(null);
        } else {
            Node<T> curr = this.getFirst().getNext();
            this.getFirst().setNext(null);
            this.setFirst(curr);
            curr = null;
        }

        dec();
        return content;
    }

    //=====PEEK=====//
    public T peek() {
        return this.getFirst().getObj();
    }

    //=====INVERT=====//
    public void invert() {
        Node<T> curr = this.getLast();

        while(curr != this.getFirst()) {
            curr.setNext(new Node<>(dequeue(), curr.getNext()));
            if(this.getLast() == curr) this.setLast(curr.getNext());
            inc();
        }

        curr = null;
    }

    //=====OVERRIDE=====//
    @Override
    public String toString() {
        StringBuffer sb = new StringBuffer("[");
        Node<T> curr = this.getFirst();
        while(curr != null) {
            sb.append(curr.getObj());
            if(curr.getNext() != null) sb.append(", ");
            curr = curr.getNext();
        }
        curr = null;
        return sb.append("]").toString();
    }

}

File: EstruturasFlexiveis/OutrasEstruturas/Queue/C/Queue.c

#include "header/Queue.h"

//=====CONSTRUCT======//
Queue* construct() {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    clear(queue);
    return queue;
}

//=====INITIALIZER=====//
void clear(Queue* const queue) {
    queue->first = NULL;
    queue->last = NULL;
    queue->len = 0;
}

//=====DESTRUCTOR=====//
void destruct(Queue* const queue) {
    Node* prev = NULL;
    Node* curr = queue->first;

    while(curr != NULL) {
        prev = curr;
        curr = curr->next;
        free(prev);
    }

    free(queue);
}

//=====METODOS=====//
//-----ISEMPTY-----//
bool isEmpty(const Queue* const queue) {
    return queue->len == 0;
}

//-----ENQUEUE-----//
void enqueue(Queue* const queue, int obj) {
    Node* node = constructNodeObj(obj);

    if(isEmpty(queue)) queue->first = queue->last = node;
    else {
        queue->last->next = node;
        queue->last = node;
    }

    queue->len++;
}

//-----DEQUEUE-----//
int dequeue(Queue* const queue) {
    Node* node = NULL;
    
    if(!isEmpty(queue)) {
        node = queue->first;
        if(queue->first == queue->last) queue->first = queue->last = NULL;
        else {
            Node* rem = queue->first;
            queue->first = queue->first->next;
            rem = rem->next = NULL;
        }
    }

    queue->len--;
    return node->obj;
}

//-----PEEK-----//
int peek(const Queue* const queue) {
    return queue->first->obj;
}

//=====PRINT=====//
void print(const Queue* const queue) {
    printf("[");

    Node* curr = queue->first;
    while(curr != NULL) {
        printf("%i", curr->obj);
        if(curr->next != NULL) printf(", ");
        curr = curr->next;
    }

    printf("]\n");
}   
